llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,435 @@
+};
+
+Register AIEWawRegRewriter::getPhysReg(const MachineOperand &MO) const {
+  Register PhysReg = MO.getReg();
+  if (PhysReg.isVirtual())
+    PhysReg = VRM->getPhys(PhysReg);
+  return PhysReg;
+}
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopMBBs =
+      AIELoopUtils::getInnermostLoopBlocks(MLI);
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  return renameMBBPhysRegs(LoopMBBs);
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(
+    const SmallVector<const MachineBasicBlock *, 4> &LoopMBBs) {
+  bool Modified = false;
+  for (const MachineBasicBlock *MBB : LoopMBBs) {
+    LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+               dbgs() << "\n");
+    SmallVector<MCPhysReg, 16> UsedPhyRegs;
+    SmallVector<MCPhysReg, 16> UsedVirtRegs;
+    SmallVector<std::pair<MCRegister, MCRegister>, 8> ReplacementRegs;
+
+    SmallSet<MCPhysReg, 16> ForbiddenPhysRegs =
+        getMBBForbiddenReplacements(MBB);
+
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> UndefPartReuse;
+
+    for (const MachineInstr &MI : *MBB) {
+
+      if (isIdentityCopy(MI))
+        continue;
+
+      for (const MachineOperand &Op : MI.defs()) {
+        Register Reg = Op.getReg();
+        if (!Reg.isVirtual())
+          continue;
+        if (VRM->hasRequiredPhys(Reg))
+          continue;
+        if (Op.isTied())
+          continue;
+        if (!TRI->isVecOrAccRegClass(MRI->getRegClass(Reg)))
+          continue;
+        //  Subregisters can reuse Constants in the high part of the
+        //  register, therefore renaming a reused physical register can
+        //  prohibit the elimination of a move of constants, thus degrading
+        //  performance if this physical register is removed
+        if (!DisableCopySubRegHandling &&
+            isUndefReserved(Op, MI, *MBB, UndefPartReuse))
+          continue;
+
+        if (isWAWCandidate(Reg, MI, UsedPhyRegs, UsedVirtRegs)) {
+          MCRegister PreSwapPhyReg = VRM->getPhys(Reg).id();
+
+          if (replaceReg(Reg, ForbiddenPhysRegs)) {
+            LLVM_DEBUG(dbgs() << MI);
+            Modified = true;
+            ReplacementRegs.push_back(
+                std::make_pair(PreSwapPhyReg, VRM->getPhys(Reg).id()));
+          }
+
+        } else {
+          // keep track of already visted physical and virtual Regs
+          UsedPhyRegs.push_back(VRM->getPhys(Reg));
+          UsedVirtRegs.push_back(Register::virtReg2Index(Reg));
+        }
```

Comment:
There's a lot of logic in that loop, I would expect something simple like
```
if (isCandidate(...)) {
  // Then do renaming
} else {
  // Keep track of defs
}
```

Would it be possible to write the code like this? I'm wondering why some early checks have an early `continue;` statement, while those in `isWAWCandidate()` do not.

Reply:
moved the renaming function call outside of the second if condition. 
this way it is clearer, that the renaming happens and only on success, i record the changes. 
Does this clear it up?

Reply:
I rewrote the whole for clarity.